check for original coords

TimOliverMaier opened the pull request:
This is my quick fix for #5760  . When coords are set manually, `length_tensor` in `set_data` has no owner. That's why it is crashing. In my fix I only check if `original_coords` is not None, meaning they were set explicitly (as far as I understand). However, I am pretty unsure if that's the way to go. Especially since there were a couple of issues in the last time concering `coords`...

TimOliverMaier committed:
check for original coords

canyon289 commented on pymc/model.py:
```diff
@@ -1179,16 +1179,17 @@ def set_data(
-                    actual=new_length,
-                    expected=old_length,
-                )
+            if original_coords is None:
```
Add a comment to explain why this was needed, otherwise looks good

canyon289 commented:
Is there a way to write a test for this to ensure whatever bug is being fixed is fixed?

michaelosthege commented:
In light of https://github.com/pymc-devs/pymc/issues/5751 we'll have to do this slightly different.

Something like this:

1. `if length_changed`
    1. `if `length_tensor.owner.inputs[0].owner is None` → Warning: You're changing the shape of a shared variable in the '{dname}' dimension which was initialized from coords. Make sure to update the corresponding coords, otherwise you'll get shape issues.
    2. `if not isinstance(length_belongs_to, SharedVariable)` → ShapeError: You're trying to resize the '{dname}' dimension which is linked to {length_belongs_to} which can not be resized.

The i) case happens with the example from #5760.

The ii) case happens with something like this:
```python
with pm.Model():
    a = pm.Normal("a", mu=[1,2,3], dims="fixed")
    m = pm.MutableData("m", [[1,2,3]], dims=("one", "fixed"))

    # This is fine because the "fixed" dim is not resized
    pm.set_data({"m": [[1,2,3], [3,4,5]]})

    # Can't work because the "fixed" dimension is linked to a constant shape:
    # Note that the new data tries to change both dimensions
    pm.set_data({"m": [[1,2], [3,4]]})
```

@TimOliverMaier you could make that change here, ideally with a test case.

It is also relevant for https://github.com/pymc-devs/pymc/issues/5751 and AFAIK @LukeLB is working on that at the moment.

TimOliverMaier commented:
Thanks for the feedback I will look into this today. Also keeping an eye on #5751

TimOliverMaier committed:
support TensorConstant as dim length in set_data

TimOliverMaier committed:
forgot pre-commit

TimOliverMaier commented:
I rearranged the if-statements a little to cluster those relevant for the `length_changed = True` case. Furthermore the code now test if `length_tensor` has an owner. If not, it assumes the coords were set customly and logs a warning if the length was changed. I looked into the PR from @LukeLB addressing issue #5751 and added an if statement checking if dimension length is stored in TensorConstant. 
Now I am looking for the best place for tests. Does anybody have a suggestion here?

TimOliverMaier committed:
added test

TimOliverMaier commented:
I added a test only to assert that it is possible to set new data with identical dimensions.  I think additional tests should wait until a possible merge with #5763.

In `models` `set_data` stored coords differently from `add_coord`. I changed the line in `set_data` to store coords as tuple as well.

TimOliverMaier committed:
pre-commit

michaelosthege commented on pymc/model.py:
```diff
@@ -1179,23 +1179,48 @@ def set_data(
+                    # stored in TensorConstant e.g by 'fixed' set to False
+
+                    warnings.warn(
+                        f"You're changing the shape of a shared variable "
```
```suggestion
                        f"You're changing the shape of a variable "
```

michaelosthege commented on pymc/model.py:
```diff
@@ -1179,23 +1179,48 @@ def set_data(
+                        f"Resizing dimension '{dname}' is impossible, because "
+                        f"a 'TensorConstant' stores its length. To be able "
+                        f"to change the dimension length, 'fixed' in "
+                        f"'model.add_coord' must be passed False."
```
```suggestion
                        f"'model.add_coord' must be set to `False`."
```

michaelosthege commented on pymc/model.py:
```diff
@@ -1179,23 +1179,48 @@ def set_data(
+                            f"Resizing dimension '{dname}' with values of length {new_length} would lead to incompatibilities, "
+                            f"because the dimension was initialized from '{length_belongs_to}' which is not a shared variable. "
+                            f"Check if the dimension was defined implicitly before the shared variable '{name}' was created, "
+                            f"for example by a model variable.",
```
```suggestion
                            f"for example by another model variable.",
```

michaelosthege commented on pymc/model.py:
```diff
@@ -1179,23 +1179,48 @@ def set_data(
+                if original_coords is not None:
+                    if new_coords is None:
+                        raise ValueError(
+                            f"The '{name}' variable already had {len(original_coords)} coord values defined for"
```
```suggestion
                            f"The '{name}' variable already had {len(original_coords)} coord values defined for "
```

michaelosthege reviewed (approved):
I really like the informative messages!

Just a few rephrasing & typo changes... I'll apply them and merge it and then @LukeLB can close the coverage gaps after rebasing #5763.

michaelosthege committed:
Rephrase and typo fixes in error/warning messages
